scipy/stats/_binned_statistic.py

```diff
@@ -331,36 +429,51 @@ def binned_statistic_dd(sample, values, statistic='mean',
     if not callable(statistic) and statistic not in known_stats:
         raise ValueError('invalid statistic %r' % (statistic,))
 
+    # `Ndim` is the number of dimensions (e.g. `2` for `binned_statistic_2d`)
+    # `Dlen` is the length of elements along each dimension.
     # This code is based on np.histogramdd
     try:
-        # Sample is an ND-array.
-        N, D = sample.shape
+        # `sample` is an ND-array.
+        Dlen, Ndim = sample.shape
     except (AttributeError, ValueError):
-        # Sample is a sequence of 1D arrays.
+        # `sample` is a sequence of 1D arrays.
         sample = np.atleast_2d(sample).T
-        N, D = sample.shape
+        Dlen, Ndim = sample.shape
 
-    nbin = np.empty(D, int)
-    edges = D * [None]
-    dedges = D * [None]
+    try:
+        # `values` is an ND-array.
+        Vdim, Vlen = values.shape
+    except (AttributeError, ValueError):
+        # Sample is a sequence of 1D arrays.
+        values = np.atleast_2d(values)
+        Vdim, Vlen = values.shape
```

Comment:
You could just call ``np.atleast_2d`` from the beginning and skip the try/except for simplicity. If the array is already 2D, it's a no-op.